Writerside/topics/Configuration-Reference.md
# Configuration Reference

## Sensor Segment `sensors`

The sensor segment is a list of the sensors being used in the system. 
It is recommended to have an equal number of sensors as there are leds.

- **VL53L0X Sensors:**
    - `type`: vl53l0x_i2c
    - `calibration`: Tripping distance in cm, supports floats
    - `xshut_pin`: XSHUT <tooltip term="GPIO">GPIO</tooltip> connection for this sensor
    - `timing_budget`: Calculation time in μS allotted to sensor, higher values result in better accuracy but may cause lag
- **GPIO Sensors**
  - `type`: gpio
  - `pin`: <tooltip term="GPIO">GPIO</tooltip> pin of the Pi to be used
  - `invert`: Invert thr HIGH/LOW values - default: false
  - `pullup`: Use the Pi's internal pullups - default: false
  - `bounce_time`: Time in seconds to counteract "bounce" - default: 0.0

Example usage:

```yaml
sensors:
 - type: vl53l0x_i2c
   calibration: 64
   xshut_pin: 21
   timing_budget: 72000
 - type: vl53l0x_i2c
   calibration: 65.2
   xshut_pin: 20
   timing_budget: 72000
```

## Main LED Segments `leds`

Currently, only one PCA9685 I2C PWM driver is supported.

- `count`: Number of LEDs for main channels
- `freq`: Operation frequency of each LED in Hertz (higher is usually better) - default: 200
- `fps_on`: LED update rate when the system is enabled  - default: 120
- `fps_off`: LED update rate when the system is disabled - default: 60

Example usage:

```yaml
leds:
  count: 2
  freq: 240
  fps_on : 60
  fps_off: 30
```

## Extra LED Channels `extra_leds`

The extra channels section is a list of each led/sensor pair

Each list element has the below options

- `channel`: PCA channel for extra light
- `ha_name`: Entity name to appear in Home Assistant
- `ha_icon`: Home Assistant Entity icon
- `ha_id`: Home Assistant Entity unique ID
- `gpio_pin`: Pi's GPIO pin for sensor
- `gpio_pullup`: Enable Pi's built-in pullup resistor - default: false
- `gpio_invert`: Invert sensor value - default: false

Example usage:
```yaml
extra_leds:
  - channel: 15
    ha_name: "Closet Lights"
    ha_icon: "mdi:lightbulb"
    ha_id: "extra1"
    gpio_pin: 11
    gpio_pullup: True
    gpio_invert: False
```

## Per-Animation Settings `animations`

### Blink Animation `blink`
- `blink_hz`: Blink rate in Hertz - default: 2

### Fade Animation `fade`
- `fade_speed_multiplier`: Fade speed multiplier - default: 0.75

### Walking Animation `walking`
- `activation_radius`: Number of lights around the activated sensor to activate

Example usage:
```yaml
animations:
  blink:
    blink_hz: 4
  fade:
    fade_speed_multiplier: 1
  walking:
    activation_radius: 1
```

## Home Assistant and Entities `home_assistant`

### MQTT Broker Settings `mqtt`

> **Tip**
> It is **never** recommended to store passwords in the plain text config file.
> Use the `USE_ENV` option to use environment variables

The username and password options can be set to `USE_ENV` to use the `BROKER_USER` and `BROKER_PASS` environment variables respectively.
The values will default to blank if the variable isn't defined.

- `host`: Broker address - default: "homeassistant.local"
- `port`: Broker's Non-SSL MQTT port - default: 1883
- `username`: Broker username - default: "USE_ENV"
- `password`: Broker password - default: "USE_ENV"
- `connection_timeout`: Maximum time to attempt connection to MQTT - default: 6

### Device Listing `device`

- `name`: Device name - default: "AutoLight Device"
- `id`: Device unique ID - default: "autolight_0"

### Main Controller Entity to Appear in Home Assistant `light_entity`

- `name`: Entity name - default: "Main Control"
- `icon`: Entity icon - default: "mdi:lightbulb"
- `id`: Light unique ID - default: "autolight_main"

### Sensor entities to appear in Home Assistant `sensor_entities`

- `naming_scheme`: Entity naming scheme to be used in Home Assistant, `{0}` will be replaced with the sensor index starting at one - default: "Sensor {0}"
- `id_scheme`: Entity unique id scheme to be used in Home Assistant, `{0}` will be replaced with the sensor index starting at zero - default: "sensor_{0}"
- `device_class`: Device class to be used in Home Assistant - default: "motion"

### Extra Entities Available for CPU and Memory Usage `debugging_entities`

- `create_debug_entities`: Enabled or not - default: true
- `update_rate`: Update speed in seconds - default: 15

Example usage:
```yaml
home_assistant:
  mqtt:
    host: "homeassistant.local"
    port: 1883
    username: "USE_ENV"
    password: "USE_ENV"
    connection_timeout: 8
  device:
    name: "Staircase Lighting"
    id: "sl"
  light_entity:
    name: "Main Control"
    icon: "mdi:stairs"
    id: "light_staircase"
  sensor_entities:
    naming_scheme: "Staircase Segment {0}"
    id_scheme: "stair_motion_{0}"
    device_class: "motion"
  debugging_entities:
    create_debug_entities: True
    update_rate: 15.0
```

## Terminal Logging Settings `logging`

Possible log level values are `TRACE`, `DEBUG`, `INFO`, `WARNING`, `ERROR`, `CRITICAL`.

- `interactive_log_level`: Logging level for interactive terminals - default: `INFO`
- `regular_log_level`: Logging level for non-interactive sessions - default: `WARNING`
- `log_file`: Optional file to send logs to - default: "logger.log"
- `file_logging`: Enable logging to file - default: false
- `rich_traceback`: Enable rich tracebacks using the rich module (only available in interactive terminals) - default: true

## Misc Settings

- `do_banner`: Enable fancy startup banner for interactive sessions - default: true

settings.py
import os
import sys
import yaml

from loguru import logger

from typing import TypedDict

class VL53L0XTypedSettings(TypedDict):
    type: str
    calibration: float
    xshut_pin: int
    timing_budget: int

class GPIOSensorTypedSettings(TypedDict):
    type: str
    pin: float
    invert: int
    pullup: int
    bounce_time: float

class LedTypedSettings(TypedDict):
    count: int
    freq: int
    fps_on: int
    fps_off: int

class _ExtraLedTypedSetting(TypedDict):
    channel: int
    ha_name: str
    ha_icon: str
    ha_id: str
    sensor: GPIOSensorTypedSettings | None


ExtraLedsTypedSettings = list[_ExtraLedTypedSetting]

class _BlinkAnimationTypedSettings(TypedDict):
    blink_hz: float

class _FadeAnimationTypedSettings(TypedDict):
    fade_speed_multiplier: float

class _WalkingAnimationTypedSettings(TypedDict):
    activation_radius: int

class AnimationTypedSettings(TypedDict):
    blink: _BlinkAnimationTypedSettings
    fade: _FadeAnimationTypedSettings
    walking: _WalkingAnimationTypedSettings

class MqttTypedSettings(TypedDict):
    host: str
    port: int
    username: str
    password: str
    connection_timeout: float

class DeviceTypedSettings(TypedDict):
    name: str
    id: str

class LightEntityTypedSettings(TypedDict):
    name: str
    icon: str
    id: str

class SensorEntitiesTypedSettings(TypedDict):
    naming_scheme: str
    id_scheme: str
    device_class: str

class DebuggingEntitiesTypedSettings(TypedDict):
    create_debug_entities: bool
    update_rate: float

class HomeAssistantTypedSettings(TypedDict):
    mqtt: MqttTypedSettings
    device: DeviceTypedSettings
    light_entity: LightEntityTypedSettings
    sensor_entities: SensorEntitiesTypedSettings
    debugging_entities: DebuggingEntitiesTypedSettings

class LoggingTypedSettings(TypedDict):
    interactive_log_level: str
    regular_log_level: str
    log_file: str
    file_logging: bool
    rich_traceback: bool

class MiscTypedSettings(TypedDict):
    do_banner: bool

class Settings:
    def __init__(self, config_file="config.yaml"):
        self.config_file = config_file
        if not os.path.exists(config_file):
            logger.critical(f"Configuration file {config_file} does not exist")
            sys.exit()
        with open(config_file) as f:
            self.root_settings: dict = yaml.load(f, yaml.SafeLoader)

        # Sensor Settings
        self.sensor_settings: list[VL53L0XTypedSettings | GPIOSensorTypedSettings] = self.root_settings.get("sensors")
        self.sensor_count = len(self.sensor_settings)

        # Led Settings
        self.led_settings: LedTypedSettings = self.root_settings.get("leds")

        self.led_count = self.led_settings.get("count")
        self.led_freq = self.led_settings.get("freq", 200)
        self.led_fps_on = self.led_settings.get("fps_on", 120)
        self.led_fps_off = self.led_settings.get("fps_off", 60)

        # Extra Led Settings
        self.extra_led_settings: ExtraLedsTypedSettings = self.root_settings.get("extra_leds", [])

        self.extra_led_count = len(self.extra_led_settings)

        # Animation Settings
        self.animation_settings: AnimationTypedSettings = self.root_settings.get("animations", {})

        # Animations/Blink
        self.blink_animation_settings = self.animation_settings.get("blink", {})
        self.blink_animation_hz = self.blink_animation_settings.get("blink_hz", 2)

        # Animation/Fade
        self.fade_animation_settings = self.animation_settings.get("fade", {})
        self.fade_animation_multiplier = self.fade_animation_settings.get("fade_speed_multiplier", 0.75)

        # Animation/Walking
        self.walking_animation_settings = self.animation_settings.get("walking", {})
        self.walking_activation_radius = self.walking_animation_settings.get("activation_radius", 1)

        # HA Settings
        self.ha_settings: HomeAssistantTypedSettings = self.root_settings.get("home_assistant", {})

        # Ha/Mqtt
        self.mqtt_settings = self.ha_settings.get("mqtt", {})

        self.mqtt_host = self.mqtt_settings.get("host", "homeassistant.local")
        self.mqtt_port = self.mqtt_settings.get("port", 1883)
        self.mqtt_user = os.environ.get("BROKER_USER", "") if self.mqtt_settings.get("username", "USE_ENV") == "USE_ENV" else self.mqtt_settings.get("username", "USE_ENV")
        self.mqtt_pass = os.environ.get("BROKER_PASS", "") if self.mqtt_settings.get("password", "USE_ENV") == "USE_ENV" else self.mqtt_settings.get("password", "USE_ENV")
        self.mqtt_timeout = self.mqtt_settings.get("connection_timeout", 6.0)

        # Ha/Device
        self.device_settings = self.ha_settings.get("device", {})

        self.device_name = self.device_settings.get("name", "AutoLight Device")
        self.device_id = self.device_settings.get("id", "autolight_0")

        # Ha/Light
        self.light_entity_settings = self.ha_settings.get("light_entity", {})

        self.light_name = self.light_entity_settings.get("name", "Main Control")
        self.light_icon = self.light_entity_settings.get("icon", "mdi:lightbulb")
        self.light_id = self.light_entity_settings.get("id", "autolight_main")

        # Ha/Sensor
        self.sensor_entity_settings = self.ha_settings.get("sensor_entities", {})

        self.sensor_naming_scheme = self.sensor_entity_settings.get("naming_scheme", "Sensor {0}")
        self.sensor_id_scheme = self.sensor_entity_settings.get("id_scheme", "sensor_{0}")
        self.sensor_device_class = self.sensor_entity_settings.get("device_class", "motion")

        # Ha/Debug
        self.debug_entity_settings = self.ha_settings.get("debugging_entities", {})

        self.create_debug_entities = self.debug_entity_settings.get("create_debug_entities", True)
        self.debug_update_rate = self.debug_entity_settings.get("update_rate", 15)

        # Logging Settings
        self.logging_settings: LoggingTypedSettings = self.root_settings.get("logging", {})

        self.interactive_log_level = self.logging_settings.get("interactive_log_level", "INFO")
        self.regular_log_level = self.logging_settings.get("regular_log_level", "WARNING")

        self.log_file_path = self.logging_settings.get("log_file", "logger.log")
        self.log_to_file = self.logging_settings.get("file_logging", False)

        self.rich_tracebacks = self.logging_settings.get("rich_traceback", True)

        # Misc Settings
        self.misc_settings: MiscTypedSettings = self.root_settings.get("misc", {})

        self.do_banner = self.misc_settings.get("do_banner", True)


